File: backend/api/license/api/public/renewal.py
"""
License Renewal API Endpoint

POST /api/licenses/renew

This endpoint handles license renewal requests from the desktop app.
It checks if the license is eligible for renewal and returns a new
license key if the subscription is active.

Flow:
1. Desktop app sends renewal request with license_id + machine_fingerprint
2. Server verifies the license exists and is renewable
3. Server checks payment status (Stripe/payment provider)
4. If valid, generates new license key with extended expiration
5. Returns new license key or renewal URL for payment

Security:
- Requires valid license_id
- Machine fingerprint must match (optional, for anti-piracy)
- Rate limited to prevent abuse
"""

from datetime import datetime, timezone, timedelta

# Import license generator (relative import within fi_license package)
from backend.api.license import LicensePayload, decode_license_key, generate_license_key
from backend.infrastructure.auth.adapters.fastapi_adapter import get_current_user
from backend.infrastructure.auth.domain.entities.user import User
from backend.utils.common.logging.logger import get_logger
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

logger = get_logger(__name__)

router = APIRouter(prefix="/licenses", tags=["licenses"])


class RenewalRequest(BaseModel):
    """License renewal request from desktop app."""

    license_id: str
    machine_fingerprint: str | None = None
    current_expires_at: str | None = None


class RenewalResponse(BaseModel):
    """License renewal response."""

    renewed: bool
    reason: str | None = None
    new_expires_at: str | None = None
    new_license_key: str | None = None
    renewal_url: str | None = None
    message: str


# In-memory license registry (replace with database in production)
# Format: { license_id: { expires_at, clinic_id, payment_status, ... } }
LICENSE_REGISTRY: dict = {}


@router.post("/renew", response_model=RenewalResponse)
async def renew_license(
    request: RenewalRequest,
    current_user: User = Depends(get_current_user),
) -> RenewalResponse:
    """
    Renew a license if eligible.

    Returns:
    - renewed=True + new_license_key if successful
    - renewed=False + renewal_url if payment required
    - renewed=False + reason if not eligible
    """
    # Multi-tenancy: Validate user has clinic assigned
    if not current_user.clinic_id:
        return RenewalResponse(
            renewed=False,
            reason="no_clinic",
            message="User has no clinic assigned. Cannot renew licenses.",
        )

    license_id = request.license_id

    # Check if license exists in registry
    license_info = LICENSE_REGISTRY.get(license_id)

    if not license_info:
        # License not in registry - could be a valid license we haven't seen
        # In production, check database here
        return RenewalResponse(
            renewed=False,
            reason="license_not_found",
            message="License not found in registry. Please contact support.",
        )

    # Multi-tenancy: Validate license belongs to user's clinic
    license_clinic_id = license_info.get("clinic_id")
    if license_clinic_id != current_user.clinic_id:
        logger.error(
            "LICENSE_RENEWAL_CLINIC_IMPERSONATION_BLOCKED",
            user_clinic_id=current_user.clinic_id,
            license_clinic_id=license_clinic_id,
            license_id=license_id,
        )
        return RenewalResponse(
            renewed=False,
            reason="access_denied",
            message="Access denied: Cannot renew licenses from other clinics.",
        )

    # Check payment status
    payment_status = license_info.get("payment_status", "unknown")

    if payment_status == "active":
        # Payment is active - generate new license key
        try:
            # Calculate new expiration (extend by subscription period)
            current_expires = datetime.fromisoformat(
                license_info.get("expires_at", "").replace("Z", "+00:00")
            )
            new_expires = max(current_expires, datetime.now(UTC)) + timedelta(days=365)

            # Generate new license key
            payload = LicensePayload(
                license_id=license_id,
                clinic_id=license_info.get("clinic_id", ""),
                clinic_name=license_info.get("clinic_name", ""),
                features=license_info.get("features", []),
                expires_at=new_expires.isoformat(),
            )

            new_license_key = generate_license_key(payload)

            # Update registry
            LICENSE_REGISTRY[license_id]["expires_at"] = new_expires.isoformat()

            return RenewalResponse(
                renewed=True,
                new_expires_at=new_expires.isoformat(),
                new_license_key=new_license_key,
                message="License renewed successfully.",
            )

        except Exception as e:
            return RenewalResponse(
                renewed=False,
                reason="generation_error",
                message=f"Failed to generate new license: {e!s}",
            )

    elif payment_status in ("expired", "cancelled"):
        # Payment expired - redirect to renewal page
        renewal_url = f"https://app.aurity.io/renew?license={license_id}"

        return RenewalResponse(
            renewed=False,
            reason="payment_required",
            renewal_url=renewal_url,
            message="Your subscription has expired. Please renew to continue.",
        )

    elif payment_status == "past_due":
        # Payment past due - show warning but allow grace period
        return RenewalResponse(
            renewed=False,
            reason="payment_past_due",
            renewal_url=f"https://app.aurity.io/billing?license={license_id}",
            message="Payment is past due. Please update your payment method.",
        )

    else:
        return RenewalResponse(
            renewed=False,
            reason="unknown_status",
            message="Unable to verify license status. Please contact support.",
        )


@router.get("/status/{license_id}")
async def get_license_status(
    license_id: str,
    current_user: User = Depends(get_current_user),
) -> dict:
    """
    Get the status of a license.

    Returns license info without generating a new key.
    """
    # Multi-tenancy: Validate user has clinic assigned
    if not current_user.clinic_id:
        raise HTTPException(
            status_code=403,
            detail="User has no clinic assigned. Cannot access license status.",
        )

    license_info = LICENSE_REGISTRY.get(license_id)

    if not license_info:
        raise HTTPException(status_code=404, detail="License not found")

    # Multi-tenancy: Validate license belongs to user's clinic
    license_clinic_id = license_info.get("clinic_id")
    if license_clinic_id != current_user.clinic_id:
        logger.error(
            "LICENSE_STATUS_CLINIC_IMPERSONATION_BLOCKED",
            user_clinic_id=current_user.clinic_id,
            license_clinic_id=license_clinic_id,
            license_id=license_id,
        )
        raise HTTPException(
            status_code=403,
            detail="Access denied: Cannot access license status from other clinics.",
        )

    # Calculate days remaining
    expires_at = license_info.get("expires_at", "")
    days_remaining = None

    if expires_at:
        try:
            expires = datetime.fromisoformat(expires_at.replace("Z", "+00:00"))
            days_remaining = (expires - datetime.now(UTC)).days
        except ValueError:
            pass

    return {
        "license_id": license_id,
        "clinic_name": license_info.get("clinic_name"),
        "expires_at": expires_at,
        "days_remaining": days_remaining,
        "payment_status": license_info.get("payment_status"),
        "features": license_info.get("features", []),
    }


@router.post("/register")
async def register_license(
    license_key: str,
    current_user: User = Depends(get_current_user),
) -> dict:
    """
    Register a license key in the renewal registry.

    Called when a license is first activated to enable future renewals.
    """
    # Multi-tenancy: Validate user has clinic assigned
    if not current_user.clinic_id:
        raise HTTPException(
            status_code=403,
            detail="User has no clinic assigned. Cannot register licenses.",
        )

    try:
        payload, _ = decode_license_key(license_key)

        # Multi-tenancy: Validate license belongs to user's clinic
        if payload.clinic_id != current_user.clinic_id:
            logger.error(
                "LICENSE_REGISTER_CLINIC_IMPERSONATION_BLOCKED",
                user_clinic_id=current_user.clinic_id,
                license_clinic_id=payload.clinic_id,
                license_id=payload.license_id,
            )
            raise HTTPException(
                status_code=403,
                detail="Access denied: Cannot register licenses from other clinics.",
            )

        # Store in registry
        LICENSE_REGISTRY[payload.license_id] = {
            "license_id": payload.license_id,
            "clinic_id": payload.clinic_id,
            "clinic_name": payload.clinic_name,
            "features": payload.features,
            "expires_at": payload.expires_at,
            "payment_status": "active",  # Default to active on registration
            "registered_at": datetime.now(UTC).isoformat(),
        }

        return {
            "registered": True,
            "license_id": payload.license_id,
            "message": "License registered for renewal service.",
        }

    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

File: apps/aurity-desktop/src-tauri/src/license/renewal.rs
use log::info;

use super::storage::{activate_license, get_license_status};
use super::types::{LicenseError, RenewalRequest, RenewalResponse, RenewalStatus};

/// Renewal API base URL (production)
const RENEWAL_API_BASE: &str = "https://app.aurity.io/api";

/// Get machine fingerprint for license binding (simplified)
fn get_machine_fingerprint() -> Option<String> {
    let hostname = hostname::get().ok()?.into_string().ok()?;

    #[cfg(target_os = "macos")]
    let os_id = "macos";
    #[cfg(target_os = "windows")]
    let os_id = "windows";
    #[cfg(target_os = "linux")]
    let os_id = "linux";
    #[cfg(not(any(target_os = "macos", target_os = "windows", target_os = "linux")))]
    let os_id = "unknown";

    use std::collections::hash_map::DefaultHasher;
    use std::hash::{Hash, Hasher};

    let mut hasher = DefaultHasher::new();
    format!("{}-{}", hostname, os_id).hash(&mut hasher);
    Some(format!("{:016x}", hasher.finish()))
}

/// Check if license needs renewal (called periodically)
pub fn check_renewal_status() -> RenewalStatus {
    let status = get_license_status();

    if !status.is_valid && status.status != "expired" {
        return RenewalStatus {
            needs_renewal: false,
            days_until_expiry: None,
            renewal_url: None,
            warning_message: Some("License not activated".to_string()),
        };
    }

    let days = status.days_remaining.unwrap_or(0);

    let warning_message = if days < 0 {
        Some("Your license has expired. Please renew to continue using Aurity.".to_string())
    } else if days <= 7 {
        Some(format!(
            "Your license expires in {} days. Please renew soon.",
            days
        ))
    } else if days <= 30 {
        Some(format!("Your license expires in {} days.", days))
    } else {
        None
    };

    let license_id = status
        .payload
        .as_ref()
        .map(|p| p.license_id.clone())
        .unwrap_or_default();

    RenewalStatus {
        needs_renewal: days <= 30,
        days_until_expiry: status.days_remaining,
        renewal_url: if days <= 30 {
            Some(format!(
                "https://app.aurity.io/renew?license={}",
                license_id
            ))
        } else {
            None
        },
        warning_message,
    }
}

/// Request license renewal from server (async HTTP call)
pub async fn request_renewal() -> Result<RenewalResponse, LicenseError> {
    let status = get_license_status();

    let payload = status.payload.ok_or(LicenseError::NotActivated("No valid license to renew".into()))?;

    let request = RenewalRequest {
        license_id: payload.license_id.clone(),
        machine_fingerprint: get_machine_fingerprint(),
        current_expires_at: Some(payload.expires_at.clone()),
    };

    let client = reqwest::Client::new();
    let response = client
        .post(format!("{}/licenses/renew", RENEWAL_API_BASE))
        .json(&request)
        .timeout(std::time::Duration::from_secs(30))
        .send()
        .await
        .map_err(|e| LicenseError::Network(format!("Network error: {}", e)))?;

    if !response.status().is_success() {
        let status = response.status();
        let body = response.text().await.unwrap_or_default();
        return Err(LicenseError::Network(format!("Server error {}: {}", status, body)));
    }

    let renewal: RenewalResponse = response
        .json()
        .await
        .map_err(|e| LicenseError::Network(format!("Invalid response: {}", e)))?;

    if renewal.renewed {
        if let Some(ref new_key) = renewal.new_license_key {
            activate_license(new_key)?;
            info!("License renewed successfully!");
        }
    }

    Ok(renewal)
}

/// Register license for renewal service
pub async fn register_for_renewal(license_key: &str) -> Result<(), LicenseError> {
    let client = reqwest::Client::new();

    let response = client
        .post(format!("{}/licenses/register", RENEWAL_API_BASE))
        .query(&[("license_key", license_key)])
        .timeout(std::time::Duration::from_secs(30))
        .send()
        .await
        .map_err(|e| LicenseError::Network(format!("Network error: {}", e)))?;

    if !response.status().is_success() {
        let status = response.status();
        let body = response.text().await.unwrap_or_default();
        return Err(LicenseError::Network(format!("Registration failed {}: {}", status, body)));
    }

    info!("License registered for renewal service");
    Ok(())
}
